util/EIP-712/src/parser.rs

```diff
@@ -0,0 +1,176 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Solidity type-name parsing
+use lunarity_lexer::{Lexer, Token};
+use error::*;
+use toolshed::Arena;
+use std::{fmt, result};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Type {
+	Address,
+	Uint,
+	Int,
+	String,
+	Bool,
+	Bytes,
+	Byte(u8),
+	Custom(String),
+	Array {
+		length: u64,
+		inner: Box<Type>
+	}
+}
+
+impl From<Type> for String {
+	fn from(field_type: Type) -> String {
+		match field_type {
+			Type::Address => "address".into(),
+			Type::Uint => "uint".into(),
+			Type::Int => "int".into(),
+			Type::String => "string".into(),
+			Type::Bool => "bool".into(),
+			Type::Bytes => "bytes".into(),
+			Type::Byte(len) => format!("bytes{}", len),
+			Type::Custom(custom) => custom,
+			Type::Array {
+				inner,
+				length
+			} => {
+				let inner: String = (*inner).into();
+				if length == 0 {
+					format!("{}[]", inner)
+				} else {
+					format!("{}[{}]", inner, length)
+				}
+			}
+		}
+	}
+}
+
+impl fmt::Display for Type {
+	fn fmt(&self, f: &mut fmt::Formatter) -> result::Result<(), fmt::Error> {
+		let item: String = self.clone().into();
+		write!(f, "{}", item)
+	}
+}
+
+pub struct Parser {
+	arena: Arena,
+}
+
+impl Parser {
+	pub fn new() -> Self {
+		Parser {
+			arena: Arena::new()
+		}
+	}
+
+	/// the type string is being validated before it's parsed.
+	pub fn parse_type(&self, field_type: &str) -> Result<Type> {
+		#[derive(PartialEq)]
+		enum State { Open, Close }
+
+		let mut lexer = Lexer::new(&self.arena, field_type);
+		let mut token = None;
+		let mut state = State::Close;
+		let mut array_depth = 0;
+		let mut current_array_length: Option<u64> = None;
+
+		while lexer.token != Token::EndOfProgram  {
+			let type_ = match lexer.token {
+				Token::Identifier => Type::Custom(lexer.token_as_str().to_owned()),
+				Token::TypeByte => Type::Byte(lexer.type_size.0),
+				Token::TypeBytes => Type::Bytes,
+				Token::TypeBool => Type::Bool,
+				Token::TypeUint => Type::Uint,
+				Token::TypeInt => Type::Int,
+				Token::TypeString => Type::String,
+				Token::TypeAddress => Type::Address,
+				Token::LiteralInteger => {
+					let length = lexer.token_as_str();
+					current_array_length = Some(length
+						.parse()
+						.map_err(|_|
+							ErrorKind::InvalidArraySize(length.into())
+						)?
+					);
+					lexer.consume();
+					continue;
+				},
+				Token::BracketOpen if token.is_some() && state == State::Close => {
+					state = State::Open;
+					lexer.consume();
+					continue
+				}
+				Token::BracketClose if array_depth < 10 => {
+					if state == State::Open && token.is_some() {
+						let length = match current_array_length.take() {
+							None => 0,
+							Some(len) => len
```

Comment:
Seems that both `Type[0]` and `Type[]` would be parsed to the same value, is that correct?

Reply:
but i think the code could be more declarative.